### PR TITLE
WIP: .eh_frame: Implement BPF stack unwinding with .eh_frame tables

### PR DESCRIPTION
**Pull request by javierhonduco**
This commit is a work-in-progress (1) of stack unwinding with the .eh_frame tables. The main goal of this code is to support the `parca-demo-cpp` binary soon (1), and build from there to support every binary and other architectures later on. We will also work on reducing the minimum kernel version required later on. See https://github.com/parca-dev/parca-agent/issues/768 for more details on the goals and non-goals of the first iteration.

## How to run

1. Run the test binary: `./internal/dwarf/frame/testdata/parca-demo-cpp-no-fp`
2. Start Parca Agent: `make && sudo  ./dist/parca-agent  --node=test --process=$(pidof parca-demo-cpp-no-fp)`

### Verifying that the BPF program is reading the right data

Useful to double-check that the data we are reading is the data we expect.

**On the agent**:
```
        - Total entries 60969

        row[0]. Loc: 401020, CFA Reg: 7 Offset:16, $rbp: 0
        row[1]. Loc: 401026, CFA Reg: 7 Offset:24, $rbp: 0
        row[2]. Loc: 401030, CFA Reg: 7 Offset:24, $rbp: 0
        row[60968]. Loc: 7f462ffd0ca0, CFA Reg: 7 Offset:8, $rbp: 0
```

The BPF debug output:

```
$ sudo cat /sys/kernel/debug/tracing/trace_pipe
[...]
 - unwind table has 60969 items
[...]
  parca-demo-cpp-318748  [006] d.h2. 17414.354239: bpf_trace_printk: ~ 0 entry. Loc: 401020, CFA reg: 7 Offset: 16, $rbp 0
  parca-demo-cpp-318748  [006] d.h2. 17414.354239: bpf_trace_printk: ~ 1 entry. Loc: 401026, CFA reg: 7 Offset: 24, $rbp 0
  parca-demo-cpp-318748  [006] d.h2. 17414.354240: bpf_trace_printk: ~ 2 entry. Loc: 401030, CFA reg: 7 Offset: 24, $rbp 0
  parca-demo-cpp-318748  [006] d.h2. 17414.354240: bpf_trace_printk: ~ 60968 entry. Loc: 7f462ffd0ca0, CFA reg: 7 Offset: 8, $rbp 0
[...]
```

## (1) What makes this PR a work-in-progress?

Unwinding doesn't work if `$rbp` has to be recalculated.
- How to properly restore `$rbp`. This is not implemented yet, so `internal/dwarf/frame/testdata/parca-demo-cpp` won't work yet.
- ~~Reading `$rip` is yielding the correct results even though the address we are reading is ok. This needs further investigation and will open an issue with more details later today.~~ I was reading the data from the wrong execution path 😅 

There are also some hacks that have been commented which I will work on later on to remove or improve them.

## Follow-ups

There's many things to do to consider this first implementation complete, beyond making it work 😄 . Everything should be documented in  https://github.com/parca-dev/parca-agent/issues/768

## Test plan

Sampled a couple of stacktraces and verified that:
- The initial data is correct (register, # of items, main PC range, etc)
- Each frame's data is ok (prev sp, prev ip, etc)

```
$ sudo cat /sys/kernel/debug/tracing/trace_pipe
 parca-demo-cpp--995709  [001] d.h2. 27441.478292: bpf_trace_printk: ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 parca-demo-cpp--995709  [001] d.h2. 27441.478293: bpf_trace_printk: traversing stack using .eh_frame information!!
 parca-demo-cpp--995709  [001] d.h2. 27441.478293: bpf_trace_printk: ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 parca-demo-cpp--995709  [001] d.h2. 27441.478294: bpf_trace_printk: - unwind table has 60957 items
 parca-demo-cpp--995709  [001] d.h2. 27441.478294: bpf_trace_printk: - main pc range 401260...4012d8
 parca-demo-cpp--995709  [001] d.h2. 27441.478295: bpf_trace_printk: ## frame: 0
 parca-demo-cpp--995709  [001] d.h2. 27441.478295: bpf_trace_printk:    current pc: 4011f6
 parca-demo-cpp--995709  [001] d.h2. 27441.478295: bpf_trace_printk:    current sp: 7ffe73552638
 parca-demo-cpp--995709  [001] d.h2. 27441.478296: bpf_trace_printk:    current bp: 1
 parca-demo-cpp--995709  [001] d.h2. 27441.478297: bpf_trace_printk:    .done
 parca-demo-cpp--995709  [001] d.h2. 27441.478297: bpf_trace_printk:    => table_index: 12
 parca-demo-cpp--995709  [001] d.h2. 27441.478297: bpf_trace_printk:    cfa reg: 7, offset: 8 (pc: 4011d0)
 parca-demo-cpp--995709  [001] d.h2. 27441.478298: bpf_trace_printk:    previous ip: 401206 (@ 7ffe73552638)
 parca-demo-cpp--995709  [001] d.h2. 27441.478299: bpf_trace_printk:    previous sp: 7ffe73552640
 parca-demo-cpp--995709  [001] d.h2. 27441.478299: bpf_trace_printk:    rbp offset 0
 parca-demo-cpp--995709  [001] d.h2. 27441.478299: bpf_trace_printk:    previous bp: 1
 parca-demo-cpp--995709  [001] d.h2. 27441.478299: bpf_trace_printk: ## frame: 1
 parca-demo-cpp--995709  [001] d.h2. 27441.478300: bpf_trace_printk:    current pc: 401206
 parca-demo-cpp--995709  [001] d.h2. 27441.478300: bpf_trace_printk:    current sp: 7ffe73552640
 parca-demo-cpp--995709  [001] d.h2. 27441.478300: bpf_trace_printk:    current bp: 1
 parca-demo-cpp--995709  [001] d.h2. 27441.478301: bpf_trace_printk:    .done
 parca-demo-cpp--995709  [001] d.h2. 27441.478301: bpf_trace_printk:    => table_index: 14
 parca-demo-cpp--995709  [001] d.h2. 27441.478301: bpf_trace_printk:    cfa reg: 7, offset: 16 (pc: 401201)
 parca-demo-cpp--995709  [001] d.h2. 27441.478302: bpf_trace_printk:    previous ip: 401216 (@ 7ffe73552648)
 parca-demo-cpp--995709  [001] d.h2. 27441.478302: bpf_trace_printk:    previous sp: 7ffe73552650
 parca-demo-cpp--995709  [001] d.h2. 27441.478302: bpf_trace_printk:    rbp offset 0
 parca-demo-cpp--995709  [001] d.h2. 27441.478302: bpf_trace_printk:    previous bp: 1
 parca-demo-cpp--995709  [001] d.h2. 27441.478303: bpf_trace_printk: ## frame: 2
 parca-demo-cpp--995709  [001] d.h2. 27441.478303: bpf_trace_printk:    current pc: 401216
 parca-demo-cpp--995709  [001] d.h2. 27441.478303: bpf_trace_printk:    current sp: 7ffe73552650
 parca-demo-cpp--995709  [001] d.h2. 27441.478303: bpf_trace_printk:    current bp: 1
 parca-demo-cpp--995709  [001] d.h2. 27441.478304: bpf_trace_printk:    .done
 parca-demo-cpp--995709  [001] d.h2. 27441.478304: bpf_trace_printk:    => table_index: 17
 parca-demo-cpp--995709  [001] d.h2. 27441.478305: bpf_trace_printk:    cfa reg: 7, offset: 16 (pc: 401211)
 parca-demo-cpp--995709  [001] d.h2. 27441.478305: bpf_trace_printk:    previous ip: 401226 (@ 7ffe73552658)
 parca-demo-cpp--995709  [001] d.h2. 27441.478305: bpf_trace_printk:    previous sp: 7ffe73552660
 parca-demo-cpp--995709  [001] d.h2. 27441.478305: bpf_trace_printk:    rbp offset 0
 parca-demo-cpp--995709  [001] d.h2. 27441.478306: bpf_trace_printk:    previous bp: 1
 parca-demo-cpp--995709  [001] d.h2. 27441.478306: bpf_trace_printk: ## frame: 3
 parca-demo-cpp--995709  [001] d.h2. 27441.478306: bpf_trace_printk:    current pc: 401226
 parca-demo-cpp--995709  [001] d.h2. 27441.478306: bpf_trace_printk:    current sp: 7ffe73552660
 parca-demo-cpp--995709  [001] d.h2. 27441.478307: bpf_trace_printk:    current bp: 1
 parca-demo-cpp--995709  [001] d.h2. 27441.478307: bpf_trace_printk:    .done
 parca-demo-cpp--995709  [001] d.h2. 27441.478307: bpf_trace_printk:    => table_index: 20
 parca-demo-cpp--995709  [001] d.h2. 27441.478308: bpf_trace_printk:    cfa reg: 7, offset: 16 (pc: 401221)
 parca-demo-cpp--995709  [001] d.h2. 27441.478308: bpf_trace_printk:    previous ip: 401299 (@ 7ffe73552668)
 parca-demo-cpp--995709  [001] d.h2. 27441.478308: bpf_trace_printk:    previous sp: 7ffe73552670
 parca-demo-cpp--995709  [001] d.h2. 27441.478308: bpf_trace_printk:    rbp offset 0
 parca-demo-cpp--995709  [001] d.h2. 27441.478309: bpf_trace_printk:    previous bp: 1
 parca-demo-cpp--995709  [001] d.h2. 27441.478309: bpf_trace_printk: ## frame: 4
 parca-demo-cpp--995709  [001] d.h2. 27441.478309: bpf_trace_printk:    current pc: 401299
 parca-demo-cpp--995709  [001] d.h2. 27441.478309: bpf_trace_printk:    current sp: 7ffe73552670
 parca-demo-cpp--995709  [001] d.h2. 27441.478310: bpf_trace_printk:    current bp: 1
 parca-demo-cpp--995709  [001] d.h2. 27441.478310: bpf_trace_printk: ======= reached main! =======
 parca-demo-cpp--995709  [001] d.h2. 27441.478311: bpf_trace_printk: ~ 0 entry. Loc: 401020, CFA reg: 7 Offset: 16, $rbp 0
 parca-demo-cpp--995709  [001] d.h2. 27441.478311: bpf_trace_printk: ~ 1 entry. Loc: 401026, CFA reg: 7 Offset: 24, $rbp 0
 parca-demo-cpp--995709  [001] d.h2. 27441.478312: bpf_trace_printk: ~ 2 entry. Loc: 401030, CFA reg: 7 Offset: 24, $rbp 0
 parca-demo-cpp--995709  [001] d.h2. 27441.478312: bpf_trace_printk: ~ 60956 entry. Loc: 7feb1bc43ca0, CFA reg: 7 Offset: 8, $rbp 0
````

Signed-off-by: Francisco Javier Honduvilla Coto <javierhonduco@gmail.com>

### CODE REVIEW

File: bpf/cpu/cpu.bpf.c [hunk 1]
```diff
@@ -181,187 +159,214 @@ struct {
...
-    if (guess == target)
-      return &mid;
-    else if (guess < target)
-      left = mid + 1;
-    else
-      right = mid - 1;
+// Context for the binary search. We mutate it on every step and at the end
+// we read the resulting values.
+struct callback_ctx {
+  u64 pc; // the needle.
```
Comment (kakkoyun) [inline bpf/cpu/cpu.bpf.c:166]: :joy:

File: pkg/profiler/cpu/cpu.go [hunk 2]
```diff
@@ -57,7 +57,7 @@ const (
 
 	defaultRLimit = 1024 << 20 // ~1GB
 
-	programName = "profile_cpu"
+	programName = "do_sample"
```
Comment (kakkoyun) [inline pkg/profiler/cpu/cpu.go:60]: :see_no_evil: I think I forgot to rename the function in bpf program while rebasing.

File: pkg/profiler/cpu/maps.go [hunk 3]
```diff
@@ -97,46 +99,66 @@ func (m *bpfMaps) readStackCount(keyBytes []byte) (uint64, error) {
 
 // updateUnwindTables updates the unwind tables with the given plan table.
 func (m bpfMaps) updateUnwindTables(pid int, pt unwind.PlanTable) error {
```
Comment (kakkoyun) [inline pkg/profiler/cpu/maps.go:101]:
```suggestion
func (m *bpfMaps) updateUnwindTables(pid int, pt unwind.PlanTable) error {
```

File: pkg/stack/unwind/plan_table.go [hunk 4]
```diff
@@ -73,15 +75,28 @@ func (ptb *PlanTableBuilder) PlanTableForPid(pid int) (PlanTable, error) {
...
+		if strings.Contains(executablePath, mappings[0].File) {
+			fmt.Println("! Start set to zero for the main binary", mappings[0].File)
+			res = append(res, buildTable(fdes, 0)...)
+		} else {
+			res = append(res, buildTable(fdes, m.Start)...)
+		}
 	}
+
+	// Sort the entries so we can binary search over them.
+	sort.Sort(res)
```
Comment (kakkoyun) [inline pkg/stack/unwind/plan_table.go:99]: :+1:

File: pkg/profiler/cpu/cpu.go (hunk 2 above)
Comment (kakkoyun) [reply pkg/profiler/cpu/cpu.go:60]: Let's rename the function in the C file.

Comment (javierhonduco) [reply pkg/profiler/cpu/cpu.go:60]: Sounds good! Perhaps a name like `take_sample` is more representative of what the function is doing as the "profile CPU" bit happens on the perf event side?

Comment (kakkoyun) [reply pkg/profiler/cpu/cpu.go]: Sounds good to me!

**Review by kakkoyun — APPROVED** [review]
Looking awesome! Great job!

We still have things to clean and edge cases to handle, but this is a great success!

**Pull request merged**